### ISSUE TITLE
FEAT: Saving intermediate images

### ISSUE THREAD

**Issue by morganavr** [lstein/stable-diffusion#154]
When I generate image with 512x768px size it generates nice image but it takes long time and I was wondering if it's possible to get more images by saving intermediate images as well to a separate folder named exactly as the image generated just without .png

**Why it's useful:**
1) It allows to optimize `steps` parameter with different samplers and different resolutions
2) It allows to see how the image evolves over time

I have created the same Issue in @basujindal fork at https://github.com/basujindal/stable-diffusion/issues/75
I believe this feature could be useful in both forks. I use @basujindal fork to generate max size images on 8GB VRAM GPU and @lstein for anything else.

**Comment by BlueAmulet** [comment]
See https://github.com/lstein/stable-diffusion/issues/99#issuecomment-1229048855 as to why this may not be what you desire. The intermediary images aren't that different from the final image, besides being more blurry and fuzzy and unrefined.

It is possible to do though, the samplers have callbacks that could be added to save the image of every Nth step.

**Comment by SMUsamaShah** [comment]
> See [#99 (comment)](https://github.com/lstein/stable-diffusion/issues/99#issuecomment-1229048855) as to why this may not be what you desire. The intermediary images aren't that different from the final image, besides being more blurry and fuzzy and unrefined.
> 
> It is possible to do though, the samplers have callbacks that could be added to save the image of every Nth step.

I have discovered that k_euler_a never converges to one image. You will see a different result even on 10,000 steps. I made and posted a video upto 500 steps on reddit /r/StableDiffusion. You can see it keeps changing till the end. I then separately run jumps upto 5000 and 10000 and it still changes.

TL;DR, intermediate results are useful in case of k_euler_a

**Comment by bakkot** [comment]
I added the ability to watch images get rendered in the webui [here](https://github.com/lstein/stable-diffusion/pull/129#issuecomment-1229542947) if you want to play with it. As @BlueAmulet says, intermediates aren't really useful. They're even less useful with other samplers, like the default `k_lms`. Here's "a dragon flying past a sunset, oil painting" with the same settings at 40-of-40 steps vs 40-of-50 steps:
![dragon-40-of-40](https://user-images.githubusercontent.com/1653598/187096224-c6689cfa-ae3a-48a6-8820-f2310de759f9.jpg)
![dragon-40-of-50](https://user-images.githubusercontent.com/1653598/187096228-755b5f33-0946-44a4-b117-371d9a9678ab.jpg)

**Comment by SMUsamaShah** [comment]
I tried before latest pull locally and was getting errors on img.save saying it wasn't. Thought i should pull before diving in further. Came to github and here you have it. Thank you :)

**Comment by BlueAmulet** [comment]
> I have discovered that k_euler_a never converges to one image.

The images in that comment are of k_euler_a. Changing the number of steps does change the resulting final image yes, but the intermediary of k_euler_a at step 50 of 100 total, does not look like the final result of k_euler_a at 50 steps total.

**Comment by SMUsamaShah** [comment]
> > I have discovered that k_euler_a never converges to one image.
> 
> The images in that comment are of k_euler_a. Changing the number of steps does change the resulting final image yes, but the intermediary of k_euler_a at step 50 of 100 total, does not look like the final result of k_euler_a at 50 steps total.

Oh I now understand what you are saying. And you are right, if that's the case than they won't be of much use.
